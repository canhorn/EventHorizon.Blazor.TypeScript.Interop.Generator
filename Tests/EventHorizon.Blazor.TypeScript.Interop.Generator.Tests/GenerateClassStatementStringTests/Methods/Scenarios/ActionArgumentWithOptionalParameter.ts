declare module Examples {
  export class ExampleClass {
    methodWithCallback(
      callback: (body: string, optional?: number) => void
    ): void;
  }
}
