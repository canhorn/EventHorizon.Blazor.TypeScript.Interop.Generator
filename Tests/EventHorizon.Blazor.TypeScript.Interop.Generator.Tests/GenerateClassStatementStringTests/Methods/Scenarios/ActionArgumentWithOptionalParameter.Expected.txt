/// Generated - Do Not Edit
namespace Examples
{
    using System;
    using System.Collections.Generic;
    using System.Text.Json.Serialization;
    using System.Threading.Tasks;
    using EventHorizon.Blazor.Interop;
    using EventHorizon.Blazor.Interop.Callbacks;
    using EventHorizon.Blazor.Interop.ResultCallbacks;
    using Microsoft.JSInterop;

    
    
    [JsonConverter(typeof(CachedEntityConverter<ExampleClass>))]
    public class ExampleClass : CachedEntityObject
    {
        #region Static Accessors

        #endregion

        #region Static Properties

        #endregion

        #region Static Methods

        #endregion

        #region Accessors

        #endregion

        #region Properties

        #endregion
        
        #region Constructor
        public ExampleClass() : base() { }

        public ExampleClass(
            ICachedEntity entity
        ) : base(entity)
        {
            ___guid = entity.___guid;
        }


        #endregion

        #region Methods
        #region methodWithCallback TODO: Get Comments as metadata identification
        private bool _isMethodWithCallbackEnabled = false;
        private readonly IDictionary<string, Func<string, System.Nullable<decimal>, Task>> _methodWithCallbackActionMap = new Dictionary<string, Func<string, System.Nullable<decimal>, Task>>();

        public string methodWithCallback(
            Func<string, System.Nullable<decimal>, Task> callback
        )
        {
            SetupMethodWithCallbackLoop();

            var handle = Guid.NewGuid().ToString();
            _methodWithCallbackActionMap.Add(
                handle,
                callback
            );

            return handle;
        }

        public bool methodWithCallback_Remove(
            string handle
        )
        {
            return _methodWithCallbackActionMap.Remove(
                handle
            );
        }

        private void SetupMethodWithCallbackLoop()
        {
            if (_isMethodWithCallbackEnabled)
            {
                return;
            }
            EventHorizonBlazorInterop.FuncCallback(
                this,
                "methodWithCallback",
                "CallMethodWithCallbackActions",
                _invokableReference
            );
            _isMethodWithCallbackEnabled = true;
        }

        [JSInvokable]
        public async Task CallMethodWithCallbackActions(string body, System.Nullable<decimal> optional)
        {
            foreach (var action in _methodWithCallbackActionMap.Values)
            {
                await action(body, optional);
            }
        }
        #endregion
        #endregion
    }
}